inc/admin-layouts.php

```diff
@@ -0,0 +1,313 @@
+<?php
+
+/**
+ * Class SiteOrigin_Panels_Admin
+ *
+ * Handles all the admin and database interactions.
+ */
+class SiteOrigin_Panels_Admin_Layouts {
+	
+	const LAYOUT_URL = 'http://layouts.siteorigin.com/';
+	
+	function __construct() {
+		// Filter all the available external layout directories.
+		add_filter( 'siteorigin_panels_external_layout_directories', array( $this, 'filter_directories' ), 8 );
+		
+		add_action( 'wp_ajax_so_panels_layouts_query', array( $this, 'action_get_prebuilt_layouts' ) );
+		add_action( 'wp_ajax_so_panels_get_layout', array( $this, 'action_get_prebuilt_layout' ) );
+		add_action( 'wp_ajax_so_panels_import_layout', array( $this, 'action_import_layout' ) );
+		add_action( 'wp_ajax_so_panels_export_layout', array( $this, 'action_export_layout' ) );
+		add_action( 'wp_ajax_so_panels_directory_enable', array( $this, 'action_directory_enable' ) );
+	}
+	
+	/**
+	 * @return SiteOrigin_Panels_Admin_Layouts
+	 */
+	public static function single() {
+		static $single;
+		return empty( $single ) ? $single = new self() : $single;
+	}
+	
+	/**
+	 * Add the main SiteOrigin layout directory
+	 */
+	public function filter_directories( $directories ){
+		if ( apply_filters( 'siteorigin_panels_layouts_directory_enabled', true ) ) {
+			$directories['siteorigin'] = array(
+				// The title of the layouts directory in the sidebar.
+				'title' => __( 'Layouts Directory', 'siteorigin-panels' ),
+				// The URL of the directory.
+				'url'   => self::LAYOUT_URL,
+				// Any additional arguments to pass to the layouts server
+				'args'  => array()
+			);
+		}
+		
+		return $directories;
+	}
+	
+	/**
+	 * Get all the layout directories.
+	 *
+	 * @return array
+	 */
+	public function get_directories(){
+		$directories = apply_filters( 'siteorigin_panels_external_layout_directories', array() );
+		if( empty( $directories ) || ! is_array( $directories ) ) {
+			$directories = array();
+		}
+		
+		return $directories;
+	}
+	
+	/**
+	 * Gets all the prebuilt layouts
+	 */
+	function action_get_prebuilt_layouts() {
+		if ( empty( $_REQUEST['_panelsnonce'] ) || ! wp_verify_nonce( $_REQUEST['_panelsnonce'], 'panels_action' ) ) {
+			wp_die();
+		}
+		
+		// Get any layouts that the current user could edit.
+		header( 'content-type: application/json' );
+		
+		$type   = ! empty( $_REQUEST['type'] ) ? $_REQUEST['type'] : 'directory-siteorigin';
+		$search = ! empty( $_REQUEST['search'] ) ? trim( strtolower( $_REQUEST['search'] ) ) : '';
+		$page   = ! empty( $_REQUEST['page'] ) ? intval( $_REQUEST['page'] ) : 1;
```

Comment:
Small semantic issue, but I think this should rather be called `$page_num` or `$page_index` or similar. When I saw this being used below I had to return here to see what it was.